### PR TITLE
Add Rails >= 5.0.beta3 JSON API params parsing

### PR DESCRIPTION
**Pull request by bf4**
per https://github.com/cerebris/jsonapi-resources/pull/642

### CODE REVIEW

File: lib/active_model_serializers/register_jsonapi_renderer.rb [hunk 1]
```diff
@@ -38,20 +64,7 @@ def serialize_jsonapi(json, options)
...
-# actionpack/lib/action_dispatch/http/parameters.rb
-parsers::DEFAULT_PARSERS[media_type] = lambda do |body|
-  data = JSON.parse(body)
-  data = { :_json => data } unless data.is_a?(Hash)
-  data.with_indifferent_access
-end
+ActiveModelSerializers::Jsonapi.install
 
 # ref https://github.com/rails/rails/pull/21496
 ActionController::Renderers.add :jsonapi do |json, options|
```
Comment (NullVoxPopuli) [inline lib/active_model_serializers/register_jsonapi_renderer.rb]: line 72 is where travis is  failing. media_type undefined

File: lib/active_model_serializers/register_jsonapi_renderer.rb [hunk 2]
```diff
@@ -38,20 +64,7 @@ def serialize_jsonapi(json, options)
   end
 end
 
-# actionpack/lib/action_dispatch/http/mime_types.rb
-Mime::Type.register ActiveModelSerializers::Jsonapi::MEDIA_TYPE, :jsonapi
-
-parsers = Rails::VERSION::MAJOR >= 5 ? ActionDispatch::Http::Parameters : ActionDispatch::ParamsParser
-media_type = Mime::Type.lookup(ActiveModelSerializers::Jsonapi::MEDIA_TYPE)
```
Comment (NullVoxPopuli) [inline lib/active_model_serializers/register_jsonapi_renderer.rb]: line 72 needs this.

**Comment by remear** [comment]
I'm handling this PR. For now, hold off on reviewing further.

File: lib/active_model_serializers/register_jsonapi_renderer.rb [hunk 3]
```diff
@@ -29,6 +29,32 @@ module ActiveModelSerializers::Jsonapi
     response: { 'CONTENT_TYPE'.freeze => MEDIA_TYPE },
     request:  { 'ACCEPT'.freeze => MEDIA_TYPE }
   }.freeze
+  def self.install
+    # actionpack/lib/action_dispatch/http/mime_types.rb
+    Mime::Type.register MEDIA_TYPE, :jsonapi
+
+    mime_type = Mime::Type.lookup(MEDIA_TYPE)
```
Comment (beauby) [inline lib/active_model_serializers/register_jsonapi_renderer.rb]: This is equivalent to `mime_type = Mime[MEDIA_TYPE]`.

File: lib/active_model_serializers/register_jsonapi_renderer.rb [hunk 4]
```diff
@@ -29,6 +29,32 @@ module ActiveModelSerializers::Jsonapi
...
+    if Rails::VERSION::MAJOR >= 5
+      parsers = ActionDispatch::Request.parameter_parsers.merge(
+        mime_type.symbol => parser
+      )
+      ActionDispatch::Request.parameter_parsers = parsers
```
Comment (beauby) [inline lib/active_model_serializers/register_jsonapi_renderer.rb]: Why not directly `merge!`?

File: lib/active_model_serializers/register_jsonapi_renderer.rb [hunk 5]
```diff
@@ -22,43 +22,58 @@
...
   MEDIA_TYPE = 'application/vnd.api+json'.freeze
   HEADERS = {
     response: { 'CONTENT_TYPE'.freeze => MEDIA_TYPE },
     request:  { 'ACCEPT'.freeze => MEDIA_TYPE }
   }.freeze
+
+  def self.install
+    # actionpack/lib/action_dispatch/http/mime_types.rb
+    Mime::Type.register MEDIA_TYPE, :jsonapi
+    mime_type = Mime::Type.lookup(MEDIA_TYPE)
```
Comment (beauby) [inline lib/active_model_serializers/register_jsonapi_renderer.rb]: Why not `Mime[:jsonapi]`?

File: lib/active_model_serializers/register_jsonapi_renderer.rb [hunk 6]
```diff
@@ -22,43 +22,58 @@
...
+
+    if Rails::VERSION::MAJOR >= 5
+      parsers = ActionDispatch::Request.parameter_parsers.merge!(
+        mime_type.symbol => parser
+      )
+      ActionDispatch::Request.parameter_parsers = parsers
```
Comment (beauby) [inline lib/active_model_serializers/register_jsonapi_renderer.rb]: No need for the assignment if you merge the hash in place.

File: lib/active_model_serializers/register_jsonapi_renderer.rb [hunk 7]
```diff
@@ -22,43 +22,58 @@
...
+    else
+      ActionDispatch::ParamsParser::DEFAULT_PARSERS[mime_type] = parser
+    end
+
+    # ref https://github.com/rails/rails/pull/21496
+    ActionController::Renderers.add :jsonapi do |json, options|
+      json = serialize_jsonapi(json, options).to_json(options) unless json.is_a?(String)
+      self.content_type ||= Mime::Type.lookup(ActiveModelSerializers::Jsonapi::MEDIA_TYPE)
```
Comment (beauby) [inline lib/active_model_serializers/register_jsonapi_renderer.rb]: Could be `self.content_type ||= Mime[:jsonapi]`.

File: test/active_model_serializers/register_jsonapi_renderer_test_isolated.rb [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+      end
+    end
+
+    def test_jsonapi_parser_not_registered
+      parsers = if Rails::VERSION::MAJOR >= 5
+                  ActionDispatch::Request.parameter_parsers
+                else
+                  ActionDispatch::ParamsParser::DEFAULT_PARSERS
+                end
+      mime_type = Mime::Type.lookup('application/vnd.api+json')
```
Comment (beauby) [inline test/active_model_serializers/register_jsonapi_renderer_test_isolated.rb]: `Mime[:jsonapi]`

File: test/action_controller/json_api/linked_test.rb [hunk 9]
```diff
@@ -39,70 +38,76 @@ def setup_post
 
           def render_resource_without_include
             setup_post
-            render jsonapi: @post
+            render json: @post
```
Comment (beauby) [inline test/action_controller/json_api/linked_test.rb]: Why not use the `jsonapi` renderer here?

Comment (remear) [inline test/action_controller/json_api/linked_test.rb]: Using it here feels like a mix of concerns. The `jsonapi` renderer is opt-in and this test doesn't directly concern the behavior of the `jsonapi` renderer.

File: lib/active_model_serializers/register_jsonapi_renderer.rb (hunk 5 above)
Comment (remear) [inline lib/active_model_serializers/register_jsonapi_renderer.rb]: At the time it seemed like this didn't work the same between Rails 4.x and 5. I went tried again this morning it works just fine.

File: test/action_controller/json_api/linked_test.rb (hunk 9 above)
Comment (beauby) [inline test/action_controller/json_api/linked_test.rb]: Fair enough. Should there be a test where the `jsonapi` renderer is used then?

Comment (remear) [inline test/action_controller/json_api/linked_test.rb]: I've been debating whether the following adequately test this:

https://github.com/rails-api/active_model_serializers/blob/bff317686c7277e97eb7eabc42078bb3c81b973b/test/active_model_serializers/register_jsonapi_renderer_test_isolated.rb#L14-L17

https://github.com/rails-api/active_model_serializers/blob/bff317686c7277e97eb7eabc42078bb3c81b973b/test/active_model_serializers/register_jsonapi_renderer_test_isolated.rb#L113-L126

Comment (beauby) [inline test/action_controller/json_api/linked_test.rb]: Right, hadn't seen that one.

**Pull request merged**